netmap/attributes.go

```diff
@@ -0,0 +1,208 @@
+// Attribute Value is:
+//   * `True` if node enters the subnet;
+//   * `False`, otherwise.
+func WriteSubnetInfo(node *NodeInfo, info NodeSubnetInfo) {
+	attrs := node.GetAttributes()
+
+	id := info.ID()
+	enters := info.Enters()
+
+	// calculate attribute key
+	key := subnetAttributeKey(id)
+
+	if refs.IsZeroSubnet(id) == enters {
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs = append(attrs[:i], attrs[i+1:]...)
+			}
+		}
+	} else {
+		var val string
+
+		if enters {
+			val = attrSubnetValEntry
+		} else {
+			val = attrSubnetValExit
+		}
+
+		presented := false
+
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs[i].SetValue(val)
+				presented = true
+			}
+		}
+
+		if !presented {
+			var attr Attribute
+
+			attr.SetKey(key)
+			attr.SetValue(val)
+
+			attrs = append(attrs, &attr)
+		}
+	}
+
+	node.SetAttributes(attrs)
+}
+
+// ErrRemoveSubnet is returned when a node needs to leave the subnet.
+var ErrRemoveSubnet = errors.New("remove subnet")
+
+// IterateSubnets iterates over all subnets the node belongs to and passes the IDs to f.
+// Handler must not be nil.
+//
+// If f returns ErrRemoveSubnet, then removes subnet entry. Breaks on any other non-nil error and returns it.
+//
+// Returns an error if any subnet attribute has wrong format.
+func IterateSubnets(node *NodeInfo, f func(refs.SubnetID) error) error {
+	attrs := node.GetAttributes()
+
+	var (
+		err     error
+		id      refs.SubnetID
+		metZero bool // if zero subnet's attribute was met in for-loop
+	)
+
+	for i := 0; i < len(attrs); i++ { // range must not be used because of attrs mutation in body
+		key := attrs[i].GetKey()
+
+		// cut subnet ID string
+		idTxt := strings.TrimPrefix(key, attrSubnetPrefix)
+		if idTxt == key {
+			// not a subnet attribute
+			continue
+		}
+
+		// check value
+		switch val := attrs[i].GetValue(); val {
+		default:
+			return fmt.Errorf("invalid attribute value: %s", val)
+		case attrSubnetValExit:
+			// node is outside the subnet
+			continue
+		case attrSubnetValEntry:
+			// required to avoid default case
+		}
+
+		// decode subnet ID
+		if err = id.UnmarshalText([]byte(idTxt)); err != nil {
+			return fmt.Errorf("invalid ID text: %w", err)
+		}
+
+		// pass ID to the handler
+		err = f(id)
+
+		isRemoveErr := errors.Is(err, ErrRemoveSubnet)
+
+		if err != nil && !isRemoveErr {
+			return err
+		}
+
+		if !metZero { // in order to not reset if has been already set
+			metZero = refs.IsZeroSubnet(&id)
+
+			if !isRemoveErr {
+				// no handler's error and non-zero subnet
+				continue
+			} else if metZero {
+				// removal error and zero subnet.
+				// we don't remove attribute of zero subnet because it means entry
+				attrs[i].SetValue(attrSubnetValExit)
+
+				continue
+			}
+		}
+
+		if isRemoveErr {
+			// removal error and non-zero subnet.
+			// we can set False or remove attribute, latter is more memory/network efficient.
```

Comment:
If we allow multiple attributes with the same key, the subnet can be zero here

Reply:
We don't allow this.